zcash_client_sqlite/src/wallet/transparent/ephemeral.rs

```diff
@@ -0,0 +1,356 @@
+            );
+            Err(SqliteClientError::EphemeralAddressReuse(
+                address_str.to_owned(),
+                Some(txid),
+            ))
+        }
+        Some(None) if reject_all_ephemeral => Err(SqliteClientError::EphemeralAddressReuse(
+            address_str.to_owned(),
+            None,
+        )),
+        _ => Ok(()),
+    }
+}
+
+/// If `address` is one of our ephemeral addresses, mark it as having an output
+/// in a transaction that we have just created. This has no effect if `address` is
+/// not one of our ephemeral addresses.
+///
+/// Returns a `SqliteClientError::EphemeralAddressReuse` error if the address was
+/// already used.
+pub(crate) fn mark_ephemeral_address_as_used<P: consensus::Parameters>(
+    wdb: &mut WalletDb<SqlTransaction<'_>, P>,
+    ephemeral_address: &TransparentAddress,
+    tx_ref: i64,
+) -> Result<(), SqliteClientError> {
+    let address_str = encode_transparent_address_p(&wdb.params, ephemeral_address);
+    ephemeral_address_check_internal(wdb, &address_str, false)?;
+
+    wdb.conn.0.execute(
+        "UPDATE ephemeral_addresses SET used_in_tx = :used_in_tx WHERE address = :address",
+        named_params![":used_in_tx": &tx_ref, ":address": address_str],
+    )?;
+    Ok(())
+}
+
+/// If `address` is one of our ephemeral addresses, mark it as having an output
+/// in the given mined transaction (which may or may not be a transaction we sent).
+///
+/// `tx_ref` must be a valid transaction reference. This call has no effect if
+/// `address` is not one of our ephemeral addresses.
+pub(crate) fn mark_ephemeral_address_as_mined<P: consensus::Parameters>(
+    wdb: &mut WalletDb<SqlTransaction<'_>, P>,
+    address: &TransparentAddress,
+    tx_ref: i64,
+) -> Result<(), SqliteClientError> {
+    let address_str = encode_transparent_address_p(&wdb.params, address);
+
+    // Figure out which transaction was mined earlier: `tx_ref`, or any existing
+    // tx referenced by `mined_in_tx` for the given address. Prefer the existing
+    // reference in case of a tie or if both transactions are unmined.
+    // This slightly reduces the chance of unnecessarily reaching the gap limit
+    // too early in some corner cases (because the earlier transaction is less
+    // likely to be unmined).
+    //
+    // The query should always return a value if `tx_ref` is valid.
+    let earlier_ref = wdb.conn.0.query_row(
+        "SELECT id_tx FROM transactions
+         LEFT OUTER JOIN ephemeral_addresses e
+         ON id_tx = e.mined_in_tx
+         WHERE id_tx = :tx_ref OR e.address = :address
+         ORDER BY mined_height ASC NULLS LAST,
+                  tx_index ASC NULLS LAST,
+                  e.mined_in_tx ASC NULLS LAST
+         LIMIT 1",
+        named_params![":tx_ref": &tx_ref, ":address": address_str],
+        |row| row.get::<_, i64>(0),
+    )?;
+
+    wdb.conn.0.execute(
+        "UPDATE ephemeral_addresses SET mined_in_tx = :mined_in_tx WHERE address = :address",
```

Comment:
Bug: if an ephemeral address is used and mined via another wallet, we will not detect it here unless we have previously reserved the same address.

Let's say wallets A and B agree that address index X is used.
- Wallet A reserves 1 ephemeral address, and gets X + 1.
- Wallet A uses X + 1 in a transaction.
- The transaction is mined.
- Wallet B detects the transaction, and detects X + 1 as theirs due to it being within the gap limit.
- 🐛 Wallet B calls this method to mark X + 1 as mined, and fails to do so because `WHERE address = :address` does not match any row.
- 💥 Wallet B reserves 1 ephemeral address, and gets X + 1.

There's another variant of the above where wallet A reserves a total of `GAP_LIMIT + 1` ephemeral addresses over some period of time, which causes wallet B to fail to even detect the latest of those addresses as ephemeral (but otherwise proceeds as above).

Add a test that reproduces this, and then fix it.

Reply:
I extended `send_multi_step_proposed_transfer` to test this case. There's a problem documented in that test that I couldn't fix for now because it effectively requires solving #1354.